src/Microsoft.ML.DataView/DataViewTypeManager.cs

```diff
@@ -0,0 +1,274 @@
+                    throw Contracts.ExceptParam(nameof(dataViewType), $"Repeated type register. The DataView type {dataViewType} " +
+                        $"has been associated with {associatedRawType} so it cannot be associated with {rawType}.");
+                }
+
+                _typeIdToDataViewTypeIdMap.Add(rawTypeId, dataViewTypeId);
+                _dataViewTypeIdToTypeIdMap.Add(dataViewTypeId, rawTypeId);
+
+                _idToDataViewTypeMap.Add(dataViewTypeId, dataViewType);
+                _idToTypeMap.Add(rawTypeId, rawType);
+            }
+        }
+
+        /// <summary>
+        /// An instance of <see cref="TypeWithAttributesId"/> represents an unique key of its <see cref="TargetType"/> and <see cref="AssociatedAttributes"/>.
+        /// </summary>
+        private class TypeWithAttributesId
+        {
+            public Type TargetType { get; }
+            public IEnumerable<Attribute> AssociatedAttributes { get; }
+
+            public TypeWithAttributesId(Type rawType, IEnumerable<Attribute> attributes)
+            {
+                TargetType = rawType;
+                AssociatedAttributes = attributes;
+            }
+
+            /// <summary>
+            /// This function computes a hashing ID from <see name="TargetType"/> and attributes attached to it.
+            /// If a type is defined as a member in a <see langword="class"/>, <see name="Attributes"/> can be obtained by calling
+            /// <see cref="MemberInfo.GetCustomAttributes(bool)"/>.
+            /// </summary>
+            public override int GetHashCode()
+            {
+                if (AssociatedAttributes == null)
+                    return TargetType.GetHashCode();
+
+                var code = TargetType.GetHashCode();
+                foreach (var attr in AssociatedAttributes)
+                    code = Hashing.CombineHash(code, attr.GetHashCode());
+                return code;
+            }
+
+            public override bool Equals(object obj)
+            {
+                if (obj is TypeWithAttributesId)
+                {
+                    var other = (TypeWithAttributesId)obj;
+                    var sameType = TargetType.Equals(other.TargetType);
+
+                    var sameAttributeConfig = true;
+
+                    if (AssociatedAttributes == null && other.AssociatedAttributes == null)
+                        sameAttributeConfig = true;
+                    else if (AssociatedAttributes == null && other.AssociatedAttributes != null)
+                        sameAttributeConfig = false;
+                    else if (AssociatedAttributes != null && other.AssociatedAttributes == null)
+                        sameAttributeConfig = false;
+                    else
+                    {
+                        var zipped = AssociatedAttributes.Zip(other.AssociatedAttributes, (attr, otherAttr) => (attr, otherAttr));
```

Comment:
We should check that `AssociatedAttributes` are the same length as well, right?

Reply:
You're right.

---
In reply to: [287164576](https://github.com/dotnet/machinelearning/pull/3263#discussion_r287164576) [](ancestors = 287164576)